Expression-scoped ignores in Python 3.8.

brandtbucher opened the pull request:
Fixes #1032. On Python 3.8, we use `end_lineno` information to scope `# type: ignore` comments to enclosing expressions. Auto-formatter users, rejoice!

The `--warn-unused-ignores` semantics should be clear from the test cases. Basically, ignores in inner scopes take precedence over ignores in enclosing scopes, and the first of multiple ignores in a scope "wins".

A few notes:
 - ~This changes the type of `Errors.ignored_lines` from a `set` to a `dict`, which maps ignored lines to the line of the comment they originated from. This structure is necessary in order to support `--warn-unused-ignores`.~
 - This adds a new slot, `Context.end_line`. It defaults to `None` in the absence of an `end_lineno` or if `Context` is not an expression.
 - `ErrorInfo.origin` now has a third member, representing the end line of the error context. It is used to determine the range of lines to search for `# type: ignore` comments. If unavailable, it defaults to the same value as the origin line. 
 - Because this uses 3.8-only AST features, a new `check-38.test` file has been created, and is hidden behind a version guard in `testcheck.py`.

ilevkivskyi commented on test-data/unit/check-38.test:
```diff
@@ -0,0 +1,21 @@
+def f(a: int): ...
+f(
+    'BAD'
+)  # type: ignore
```
I would add few more tests with expressions that are not function calls. For example list and dictionary displays, operators (like + and *), etc.

ilevkivskyi commented on mypy/treetransform.py:
```diff
@@ -60,7 +60,7 @@ def __init__(self) -> None:
         # NOTE: The 'names' and 'imports' instance variables will be empty!
         new = MypyFile(self.statements(node.defs), [], node.is_bom,
-                       ignored_lines=set(node.ignored_lines))
+                       ignored_lines=node.ignored_lines)
```
I think this should make a copy instead of passing the original reference.

ilevkivskyi commented on mypy/test/testcheck.py:
```diff
@@ -85,6 +85,9 @@
     'check-newsemanal.test',
 ]
 
+if (3, 8) < sys.version_info:
```
I would flip this.
```suggestion
if  sys.version_info >= (3, 8):
```
Also please add explanatory comment.

ilevkivskyi commented on mypy/errors.py:
```diff
@@ -114,7 +114,7 @@ class Errors:
 
     # Ignore errors on these lines of each file.
-    ignored_lines = None  # type: Dict[str, Set[int]]
+    ignored_lines = None  # type: Dict[str, Dict[int, int]]
```
Please document the meaning of this dictionary.

ilevkivskyi commented on mypy/errors.py:
```diff
@@ -300,7 +300,7 @@ def clear_errors_in_targets(self, path: str, targets: Set[str]) -> None:
 
     def generate_unused_ignore_notes(self, file: str) -> None:
-        ignored_lines = self.ignored_lines[file]
+        ignored_lines = set(self.ignored_lines[file].values())
```
Add a comment reminding that this is a set of actual `# type: ignore`s in the file.

ilevkivskyi commented on mypy/nodes.py:
```diff
@@ -227,7 +227,7 @@ class MypyFile(SymbolNode):
     imports = None  # type: List[ImportBase]
     # Lines to ignore when checking
-    ignored_lines = None  # type: Set[int]
+    ignored_lines = None  # type: Dict[int, int]
```
Please add a documenting comment also here.

ilevkivskyi commented on mypy/fastparse.py:
```diff
@@ -251,7 +251,7 @@ def __init__(self,
         self.errors = errors
 
-        self.extra_type_ignores = []  # type: List[int]
+        self.type_ignores = {}  # type: Dict[int, int]
```
Also a comment here would be helpful (the one reading the code should not guess what is the direction of the mapping).

ilevkivskyi commented on mypy/fastparse.py:
```diff
@@ -271,7 +271,20 @@ def visit(self, node: Optional[AST]) -> Any:
+        node_lines = range(node.lineno, end_lineno + 1)
+        for line in node_lines:
+            if line in self.type_ignores.values():
+                self.type_ignores = {**dict.fromkeys(node_lines, line), **self.type_ignores}
```
@msullivan does mypyc support this?

brandtbucher replied:
This has been refactored not to use unpacking, either way!

ilevkivskyi commented on mypy/fastparse.py:
```diff
@@ -271,7 +271,20 @@ def visit(self, node: Optional[AST]) -> Any:
+            return
+        node_lines = range(node.lineno, end_lineno + 1)
+        for line in node_lines:
+            if line in self.type_ignores.values():
```
I am not sure how efficient is this. I think it would be better to assign `mod.type_ignores` to an attribute on `self` and use it for this check instead of recomputing this same set every time.

Also please add a comment with an example explaining what is going on here.

ilevkivskyi commented:
Please let me know when this is ready for another review.

brandtbucher commented:
Thanks for the feedback @ilevkivskyi! I've gone ahead and addressed your concerns, and am ready for another review.

I've chosen to maintain a set of source comment lines as you suggested, in addition to the mapping of ignore scopes. One nice benefit is, with a change in the order of traversal, we can now avoid expensive dictionary copies at each ignored expression node.

ilevkivskyi commented on mypy/fastparse.py:
```diff
@@ -394,13 +414,13 @@ def translate_module_id(self, id: str) -> str:
         body = self.fix_function_overloads(self.translate_stmt_list(mod.body))
-        ignores = [ti.lineno for ti in mod.type_ignores]
-        ignores.extend(self.extra_type_ignores)
+        self.type_ignores.update({line: line for line in self.type_ignore_comments})
```
I think it is less error-prone to reset `type_ignores` when visiting each file. I.e. here I would just use `self.type_ignores = {line: line for ...}`.

ilevkivskyi commented on test-data/unit/check-38.test:
```diff
@@ -0,0 +1,82 @@
+    f(
+        'BAD'
+    )  # type: ignore
+)  # type: ignore  # N: unused 'type: ignore' comment
```
I actually meant more tests without `--warn-unused-ignores`. I would keep only couple dedicated tests for `--warn-unused-ignores`, and remove it from the rest of the test cases.

ilevkivskyi commented on mypy/fastparse.py:
```diff
@@ -271,8 +276,23 @@ def visit(self, node: Optional[AST]) -> Any:
+        # If so, take the first one and grow its scope to cover the whole node:
+        for line in node_lines:
+            if line in self.type_ignore_comments:
+                self.type_ignores.update(dict.fromkeys(node_lines, line))
```
It seems to me now there is a flaw in this logic. Imagine this test case:
```python
def func(x: int) -> None: ...
lst = [func(1),
       func('no'),  # type: ignore
       func('yes'),
       func(2)]
```
I think only the error for `func('no')` should be ignored, while we still should show the error for `func('yes')`.
It looks like the actual solution should be a bit more complex.

In some sense the logic should be reversed, for every error you should check whether there is a `# type: ignore` comment in the span of error context expression for that particular error.

I am not sure what is the best way, but maybe we would need to copy the end position for every expression in the generated mypy AST.

brandtbucher replied:
I agree now, when you put it like that. Another clear example would be:

```py
def f(a: int): ...
[f("error"), f(
    "ignored"  # type: ignore
    ), f("error")]
```

...which isn’t really possible by predefining the line scope of each ignore. Really, it’s expressions that have scopes, not the ignores. This fix is exactly backwards.

I think that having the expression context grow an `end_lineno` slot and checking for ignores when reporting is both correct and more efficient. I’ll back out the current implementation and shift to that approach, if that seems right to you @ilevkivskyi!

ilevkivskyi replied:
OK, go ahead, and let me know when the PR is ready for another review.

brandtbucher commented:
I'm ready for a new review @ilevkivskyi.

I've updated my implementation to use a new `Context.end_line` slot. It's much cleaner than my first attempt, and works as expected. I've also updated the tests, as you requested.

brandtbucher commented:
Thanks for the guidance @ilevkivskyi! Happy to contribute.
